core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java

```diff
@@ -101,6 +102,7 @@
     private final Map<NodeRef<Parameter>, Expression> parameters;
     private String updateType;
     private Optional<UpdateTarget> target = Optional.empty();
+    private List<Column> targetColumns = new ArrayList<>();
```

Comment:
Please don't initialize it to empty array.  Please make it `@Nullable`, so it is clear when it is  set to empty list of columns and it was never yet set.

Reply:
And do we have to propagate this `Nullable` to `Output` and `OutputMetadata` ? Won't it be an anti pattern like we having it as `Optional<List<Column>>` ?

Reply:
`Optional<List<Column>>` is not an anti-pattern when `Optional.empty()` has different semantic than `Optional.of(List.of())`. I think in `OutputMetadata` you can simply `requireNonNull`.

Reply:
> I think in OutputMetadata you can simply requireNonNull

I think it might fail as for `visitAnalyze` and for `visitDelete` we won't be setting any column details.

Reply:
`Optional<List<Column>>` then?

Reply:
If there is actually a semantic difference between Optional.empty and Optional.of(empty List) then it is fine. 
I mean it is not fine but less not-fine than the other way so let it stay. BTW Nullable List is the same antipattern as optional list.

Reply:
@kokosing Is there actually a semantic difference between empty list and no list at all here? Is there a behavior change anywhere based on that information?

Reply:
Yes. One you don't know columns (it was not yet set, possibly bug) or a list of columns in not applicable. Other use case is when information is known and we know that there are no columns.

Reply:
Is there an explicit `if` or any other way of accessing this information in this PR? I don't see it being used anywhere

Reply:
> Other use case is when information is known and we know that there are no columns.

This looks like a valid use case but I guess we don't use it now like for rename table or set role for tables kind of queries we might have an empty list of column. But currently we don't capture those events in `QueryOutputMetadata`.  Plus instead of identifying based on the content of the `Optional<List<Column>>` we can also identify it based on the `updateType` of the query.. so now we might be able to identify this 

> One you don't know columns (it was not yet set, possibly bug) or a list of columns in not applicable

Based on the content of the list and the `updateType` /`QueryType` for a given query. In this case the semantic difference between Optional.empty and Optional.of(empty List) is to thin.

Reply:
Ok, I am not going to block you on this, it is not a place for philosophical discussions. Do as you please.
All I am trying to say is that empty optional and empty list are not distinct to the reader. If there is indeed a distinction between those two states then we should name them. So if the empty list means no columns and empty optional means, for example, "not loaded yet" then the boolean field `loaded` should be introduced so that the reader knows what is going on.